File: Self-Guided Learning/python/day_04/examples/shape.py
"""This module contains classes that represent different shapes."""

import random


class Shape:
    """
    Represents a generic shape.
    """

    def area(self):
        """
        Calculates the area of the shape.
        """
        print(
            "Every shape has an area. But each shape has a different formula to calculate it."
        )

    def __str__(self) -> str:
        return f"{self.__class__.__name__ }: {self.area()} area"


class Circle(Shape):
    """
    Represents a circle shape.

    Attributes:
        radius (float): The radius of the circle.

    Methods:
        circumference(): Calculates the circumference of the circle.
        area(): Calculates the area of the circle.
    """

    def __init__(self, radius=1):
        self.radius = radius

    def circumference(self):
        """
        Calculates the circumference of the circle.

        Returns:
            float: The circumference of the circle.
        """
        return 3.14 * 2 * self.radius

    def area(self):
        """
        Calculates the area of the circle.

        Returns:
            float: The area of the circle.
        """
        return 3.14 * self.radius**2


class Rectangle(Shape):
    """
    Represents a rectangle shape.

    Attributes:
        width (int): The width of the rectangle.
        height (int): The height of the rectangle.
    """

    def __init__(self, width=1, height=1):
        self.width = width
        self.height = height

    def perimeter(self):
        """
        Calculates the perimeter of the rectangle.

        Returns:
            int: The perimeter of the rectangle.
        """
        return 2 * (self.width + self.height)

    # NOTE: Overriding the area method from the Shape class
    def area(self):
        """
        Calculates the area of the rectangle.

        Returns:
            int: The area of the rectangle.
        """
        return self.width * self.height


class Square(Rectangle):
    """
    Represents a square shape.

    Attributes:
        side (float): The length of the square's side.

    Methods:
        perimeter(): Calculates the perimeter of the square.
        area(): Calculates the area of the square.
    """

    def __init__(self, side=1):
        super().__init__(side, side)

    def perimeter(self):
        """
        Calculates the perimeter of the square.

        Returns:
            float: The perimeter of the square.
        """
        return 4 * self.width

    def area(self):
        """
        Calculates the area of the square.

        Returns:
            float: The area of the square.
        """
        return self.width**2


def create_random_shapes(n: int):
    """
    Creates a list of randomly generated shapes.

    Parameters:
        n (int): The number of shapes to generate.

    Returns:
        list: A list of randomly generated shapes.
    """
    shapes = []

    for _ in range(n):
        shape = random.choice([Circle, Rectangle, Square])

        if isinstance(shape, Rectangle):
            shapes.append(shape(random.randint(1, 10), random.randint(1, 10)))
        else:
            shapes.append(shape(random.randint(1, 10)))
    return shapes


def main():
    """
    This is the main function that demonstrates the usage of the Circle, Square and Rectangle classes.
    It creates instances of Circle, Square and Rectangle, calculates their area and circumference (or perimeter),
    creates a list of random shapes, sorts them by area in descending order, and prints the sorted list.
    """
    circle = Circle()
    print(circle.area())
    print(circle.circumference())

    rectangle = Rectangle()
    print(rectangle.area())
    print(rectangle.perimeter())

    shapes = create_random_shapes(5)
    shapes.sort(key=lambda x: x.area(), reverse=True)

    print("Shapes sorted by area:")
    for shape in shapes:
        print(shape)


if __name__ == "__main__":
    main()

File: Samples and Concepts/HTML, CSS and JavaScript/javascript/little_lab/src/shape.js
"use strict"; // This line enables strict mode in JavaScript. That means the code will be executed in a "strict" operating context which helps catch common coding mistakes and "unsafe" actions.

/**
 * This module contains classes that represent different shapes.
 *
 * Ported from: Self-Guided Learning/python/day_04/examples/shape.py
 */

const PI = 3.14;

class Shape {
  /**
   * Calculates the area of the shape.
   *
   * In the Python example, the base implementation prints a message.
   */
  area() {
    console.log(
      "Every shape has an area. But each shape has a different formula to calculate it."
    );
    return NaN;
  }

  toString() {
    return `${this.constructor.name}: ${this.area()} area`;
  }
}

class Circle extends Shape {
  /**
   * @param {number} [radius=1]
   */
  constructor(radius = 1) {
    super();
    this.radius = radius;
  }

  /**
   * @returns {number}
   */
  circumference() {
    return PI * 2 * this.radius;
  }

  /**
   * @returns {number}
   */
  area() {
    return PI * this.radius ** 2;
  }
}

class Rectangle extends Shape {
  /**
   * @param {number} [width=1]
   * @param {number} [height=1]
   */
  constructor(width = 1, height = 1) {
    super();
    this.width = width;
    this.height = height;
  }

  /**
   * @returns {number}
   */
  perimeter() {
    return 2 * (this.width + this.height);
  }

  /**
   * @returns {number}
   */
  area() {
    return this.width * this.height;
  }
}

class Square extends Rectangle {
  /**
   * @param {number} [side=1]
   */
  constructor(side = 1) {
    super(side, side);
  }

  /**
   * @returns {number}
   */
  perimeter() {
    return 4 * this.width;
  }

  /**
   * @returns {number}
   */
  area() {
    return this.width ** 2;
  }
}

/**
 * @param {number} min
 * @param {number} max
 * @returns {number} integer in [min, max]
 */
function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

/**
 * Creates a list of randomly generated shapes.
 *
 * @param {number} n
 * @returns {Array<Shape>}
 */
function createRandomShapes(n) {
  const shapes = [];
  const constructors = [Circle, Rectangle, Square];

  for (let i = 0; i < n; i += 1) {
    const Ctor = constructors[randomInt(0, constructors.length - 1)];

    if (Ctor === Rectangle) {
      shapes.push(new Rectangle(randomInt(1, 10), randomInt(1, 10)));
    } else if (Ctor === Square) {
      shapes.push(new Square(randomInt(1, 10)));
    } else {
      shapes.push(new Circle(randomInt(1, 10)));
    }
  }

  return shapes;
}

function main() {
  const circle = new Circle();
  console.log(circle.area());
  console.log(circle.circumference());

  const rectangle = new Rectangle();
  console.log(rectangle.area());
  console.log(rectangle.perimeter());

  const shapes = createRandomShapes(5);
  shapes.sort((a, b) => b.area() - a.area());

  console.log("Shapes sorted by area:");
  for (const shape of shapes) {
    console.log(String(shape));
  }
}

module.exports = {
  Shape,
  Circle,
  Rectangle,
  Square,
  createRandomShapes,
  main,
};

if (require.main === module) {
  main();
}
